File: data_structure/tree/src/TreeDemo.py
# Python
# データ構造: 木 (Tree)

class NodeData:
    def __init__(self, value):
        self._value = value
        self._parent = None
        self._children = []
    
    def get_value(self):
        return self._value

    def get_parent(self):
        return self._parent

    def get_children(self):
        return self._children

    def set_parent(self, parent):
        self._parent = parent
        return True

    def add_child(self, child):
        child.set_parent(self)
        self._children.append(child)
        return True
    
    def remove_child(self, child):
        if child in self._children:
            child.set_parent(None)
            self._children.remove(child)
            return True
        else:
            return False
    
    def is_leaf(self):
        return len(self._children) == 0

class TreeData:
    def __init__(self):
        self._data = None

    def get(self):
        return self._data
    
    def get_height(self, node=None):
        if node is None:
            node = self._data
        if node is None:
            return 0
        if node.is_leaf():
            return 1
        return 1 + max(self.get_height(child) for child in node.get_children())
    
    def get_parent(self, node):
        return node.get_parent()
    
    def get_children(self, node):
        return node.get_children()

    def get_node(self, value, node=None):
        if node is None:
            node = self._data
        if node is None:
            return None
        
        if node.get_value() == value:
            return node
        
        for child in node.get_children():
            result = self.get_node(value, child)
            if result is not None:
                return result
        
        return None

    def add(self, parent, value):
        new_node = NodeData(value)
        if parent is None:
            # 親が指定されていない場合は、ルートとして追加
            if self._data is None:
                self._data = new_node
                return True
            else:
                # ルートが既に存在する場合はエラー
                print(f"ERROR: {value} 重複です")
                return False
        else:
            # 親ノードに子として追加
            parent.add_child(new_node)
            return True

    def remove(self, node):
        if node is None:
            return False
        
        if node == self._data:
            self._data = None
            return True
        
        parent = node.get_parent()
        if parent is not None:
            return parent.remove_child(node)
        
        return False
    
    def traverse(self, node=None, mode="pre-order"):
        if node is None:
            node = self._data
        if node is None:
            return []
        
        result = []
        
        if mode == "pre-order":
            # 先に親、次に子
            result.append(node.get_value())
            for child in node.get_children():
                result.extend(self.traverse(child, mode))
        elif mode == "post-order":
            # 先に子、次に親
            for child in node.get_children():
                result.extend(self.traverse(child, mode))
            result.append(node.get_value())
        elif mode == "level-order":
            # 階層順
            queue = [node]
            while queue:
                current = queue.pop(0)
                result.append(current.get_value())
                queue.extend(current.get_children())
        
        return result
    
    def is_leaf(self, node):
        return node is not None and node.is_leaf()
    
    def is_empty(self):
        return self._data is None
    
    def size(self, node=None):
        if node is None:
            node = self._data
        if node is None:
            return 0
        
        count = 1  # 自分自身
        for child in node.get_children():
            count += self.size(child)
        
        return count
    
    def clear(self):
        self._data = None
        return True

    def display(self):
        if self._data is None:
            return []
        
        return self.traverse(mode="level-order")

def main():
    print("Tree TEST -----> start")

    print("\nnew")
    tree_data = TreeData()
    print(f"  現在のデータ: {tree_data.display()}")

    print("\nis_empty")
    output = tree_data.is_empty()
    print(f"  出力値: {output}")

    print("\nsize")
    output = tree_data.size()
    print(f"  出力値: {output}")

    print("\nadd")
    input_params = (None, 'Root')
    print(f"  入力値: {input_params}")
    output = tree_data.add(*input_params)
    print(f"  出力値: {output}")
    print(f"  現在のデータ: {tree_data.display()}")
    
    root_node = tree_data.get()

    print("\nadd")
    input_params = (root_node, 'Child1')
    print(f"  入力値: {input_params}")
    output = tree_data.add(*input_params)
    print(f"  出力値: {output}")
    print(f"  現在のデータ: {tree_data.display()}")

    print("\nadd")
    input_params = (root_node, 'Child2')
    print(f"  入力値: {input_params}")
    output = tree_data.add(*input_params)
    print(f"  出力値: {output}")
    print(f"  現在のデータ: {tree_data.display()}")

    print("\nget_node")
    input_value = 'Child1'
    print(f"  入力値: {input_value}")
    output = tree_data.get_node(input_value)
    print(f"  出力値: {output}")
    print(f"  現在のデータ: {tree_data.display()}")

    print("\ntraverse")
    input_mode = 'pre-order'
    print(f"  入力値: {input_mode}")
    output = tree_data.traverse(mode=input_mode)
    print(f"  出力値: {output}")
    print(f"  現在のデータ: {tree_data.display()}")

    print("\nTree TEST <----- end")

if __name__ == "__main__":
    main()

File: data_structure/tree/src/TreeDemo.rb
# Ruby
# データ構造: 木 (Tree)

class NodeData
  def initialize(value)
    @value = value
    @parent = nil
    @children = []
  end
  
  def get_value
    @value
  end

  def get_parent
    @parent
  end

  def get_children
    @children
  end

  def set_parent(parent)
    @parent = parent
    true
  end

  def add_child(child)
    child.set_parent(self)
    @children << child
    true
  end
  
  def remove_child(child)
    if @children.include?(child)
      child.set_parent(nil)
      @children.delete(child)
      true
    else
      false
    end
  end
  
  def is_leaf
    @children.empty?
  end
end

class TreeData
  def initialize
    @data = nil
  end

  def get
    @data
  end
  
  def get_height(node = nil)
    node ||= @data
    return 0 if node.nil?
    return 1 if node.is_leaf
    1 + node.get_children.map { |child| get_height(child) }.max
  end
  
  def get_parent(node)
    node.get_parent
  end
  
  def get_children(node)
    node.get_children
  end

  def get_node(value, node = nil)
    node ||= @data
    return nil if node.nil?
    
    return node if node.get_value == value
    
    node.get_children.each do |child|
      result = get_node(value, child)
      return result if result
    end
    
    nil
  end

  def add(parent, value)
    new_node = NodeData.new(value)
    if parent.nil?
      if @data.nil?
        @data = new_node
        true
      else
        puts "ERROR: #{value} 重複です"
        false
      end
    else
      parent.add_child(new_node)
      true
    end
  end

  def remove(node)
    return false if node.nil?
    
    if node == @data
      @data = nil
      return true
    end
    
    parent = node.get_parent
    parent&.remove_child(node)
  end
  
  def traverse(node = nil, mode = "pre-order")
    node ||= @data
    return [] if node.nil?
    
    result = []
    
    case mode
    when "pre-order"
      result << node.get_value
      node.get_children.each do |child|
        result.concat(traverse(child, mode))
      end
    when "post-order"
      node.get_children.each do |child|
        result.concat(traverse(child, mode))
      end
      result << node.get_value
    when "level-order"
      queue = [node]
      until queue.empty?
        current = queue.shift
        result << current.get_value
        queue.concat(current.get_children)
      end
    end
    
    result
  end
  
  def is_leaf(node)
    !node.nil? && node.is_leaf
  end
  
  def is_empty
    @data.nil?
  end
  
  def size(node = nil)
    node ||= @data
    return 0 if node.nil?
    
    count = 1
    node.get_children.each do |child|
      count += size(child)
    end
    
    count
  end
  
  def clear
    @data = nil
    true
  end

  def display
    return [] if @data.nil?
    
    traverse(mode: "level-order")
  end
end

def main
  puts "Tree TEST -----> start"

  puts "\nnew"
  tree_data = TreeData.new
  puts "  現在のデータ: #{tree_data.display}"

  puts "\nis_empty"
  output = tree_data.is_empty
  puts "  出力値: #{output}"

  puts "\nsize"
  output = tree_data.size
  puts "  出力値: #{output}"

  puts "\nadd"
  input_params = [nil, 'Root']
  puts "  入力値: #{input_params}"
  output = tree_data.add(*input_params)
  puts "  出力値: #{output}"
  puts "  現在のデータ: #{tree_data.display}"
  
  root_node = tree_data.get

  puts "\nadd"
  input_params = [root_node, 'Child1']
  puts "  入力値: #{input_params}"
  output = tree_data.add(*input_params)
  puts "  出力値: #{output}"
  puts "  現在のデータ: #{tree_data.display}"

  puts "\nadd"
  input_params = [root_node, 'Child2']
  puts "  入力値: #{input_params}"
  output = tree_data.add(*input_params)
  puts "  出力値: #{output}"
  puts "  現在のデータ: #{tree_data.display}"

  puts "\nget_node"
  input_value = 'Child1'
  puts "  入力値: #{input_value}"
  output = tree_data.get_node(input_value)
  puts "  出力値: #{output}"
  puts "  現在のデータ: #{tree_data.display}"

  puts "\ntraverse"
  input_mode = 'pre-order'
  puts "  入力値: #{input_mode}"
  output = tree_data.traverse(mode: input_mode)
  puts "  出力値: #{output}"
  puts "  現在のデータ: #{tree_data.display}"

  puts "\nTree TEST <----- end"
end

main if __FILE__ == $PROGRAM_NAME